### DISCUSSION TITLE
No Content-Length header passed when getServerSideProps is used

### DISCUSSION THREAD

**Discussion by fjanevski** [vercel/next.js#39546] [Help]
I have noticed that when we use `getServerSideProps` the `Content-Length` is not calculated and because of this we are not able to use the automatic compression from AWS CloudFront (we have also set `compress: false` in `next.config.js`)

The decision not to include the header is done here: https://github.com/vercel/next.js/blob/20486c159d8538a337da6b07b0b4490a3a0d6b91/packages/next/server/send-payload/index.ts#L58 and it has to do with the response being "dynamic". But as for now, I don't understand why is Next deciding not to include this header all the time. 

Is someone else experiencing the same problem?

**Comment by VinayakHegde** [comment]
Any update on this?

**Comment by przbab** [comment]
I encountered the same behaviour. We managed to work around it by downgrading react to version 17. Version 18 sets the `__NEXT_REACT_ROOT` environment variable that ultimately results in no `Content-Length`.

Downgrading react is definitely not my preferred solution. I would love if someone more knowledgeable could say if this is a bug or desired behaviour and how to get `Content-Length` header working with react 18.

I can set up repo to reproduce the issue, let me know if you need it.

Partial trace of the issue:
`isSupportedDocument` is set to true here:
https://github.com/vercel/next.js/blob/7b7fae5a8c71770aa9b4688a0fa3b74ee6e2ea29/packages/next/server/base-server.ts#L1141-L1146
It allows for `supportsDynamicHTML` to be true
https://github.com/vercel/next.js/blob/7b7fae5a8c71770aa9b4688a0fa3b74ee6e2ea29/packages/next/server/base-server.ts#L1153-L1159
That will cause the result to be `ReadableStream`
https://github.com/vercel/next.js/blob/7b7fae5a8c71770aa9b4688a0fa3b74ee6e2ea29/packages/next/server/base-server.ts#L1348-L1354
That will land us in
https://github.com/vercel/next.js/blob/7b7fae5a8c71770aa9b4688a0fa3b74ee6e2ea29/packages/next/server/web-server.ts#L418-L427
Because `isDynamic` check looks like this
https://github.com/vercel/next.js/blob/7b7fae5a8c71770aa9b4688a0fa3b74ee6e2ea29/packages/next/server/render-result.ts#L69-L71

**Comment by KODerFunk** [comment]
In my project, a custom-server is used to implement subtle mechanics, such as complex full-page caching, and when upgrading, I also encountered a similar problem. 
Since we call Next programmatically from the level of our server, we have the opportunity to replace the `req.headers.['user-agent']` to `Googlebot` before call `app.getRequestHandler()(req, res)`, as a result of which Next began to generate a full string response. 
I've isolated the application of this hack, but haven't done any regression testing yet, and I'm not sure if it will work with a further planned upgrade to Next 13.x and above.

**Reply by KODerFunk** [reply]
In our large project, a regression was carried out - **everything is fine.**

I also spent extra time reviewing the code for both our version (12.5.1) and the current master branch content. There is a possibility that there may be problems at the **middlewares** level, but we do not use them. Whoever had a similar problem and found this discussion, this solution will also help with a 99% chance.

**Comment by przbab** [comment]
Apparently this was a mistake, it was meant to work like this only in app directory beta but landed in pages too. It was reverted in version 13.2 https://github.com/vercel/next.js/pull/45895. There was another issue connected to this one https://github.com/vercel/next.js/discussions/38606. We had the same issue but are now running 13.2.4 in production and content-length is there.

**Comment by 24jr** [comment]
Im having similar issue `RequestContentLengthMismatchError: Request body length does not match content-length header code: 'UND_ERR_REQ_CONTENT_LENGTH_MISMATCH'` in sveltekit 

I created a sveltekit server api route say example.com/api/test I can successfully call it from thunderclient in vs code but when I call that POST endpoint from my other site otherexample.com/api/test which is also a +server.js sveltekit endpoint. I'm getting same error likely from something vercel is doing with the content length. GET requests work fine maybe because I'm not passing a body.

I'm on the latest stuff sveltekit 2. I first noticed this in july 2023 and put it on hold hoping it would bee solved when I tried again months later but I'm still getting it as of Jan 1, 2024.
